### PR TITLE
Add support for pungi arguments

### PR DESCRIPTION
**Pull request by mopsfelder**
This series of patches add support to specify pungi arguments during the build of the iso image.  It also handles invalid git repositories properly, in case a non-git repository exists at the repository path.

### CODE REVIEW

File: lib/repository.py [hunk 1]
```diff
@@ -48,7 +49,7 @@ def get_git_repository(remote_repo_url, parent_dir_path):
     name = os.path.basename(os.path.splitext(url_parts.path)[0])
 
     repo_path = os.path.join(parent_dir_path, name)
-    if os.path.exists(repo_path):
+    try:
```
Comment (ghost) [inline lib/repository.py]: Please leave only the GitRepository creation line in the try block to make it clear where the exception can be raised. This makes it easier to understand the code. The code that follows it could be moved either to the `else` clause or to outside the `try`statement. See my suggestion in the last comment.

File: lib/repository.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 import os
 import urlparse
 import utils
+import shutil
```
Comment (ghost) [inline lib/repository.py]: This commit is not related to this PR. I'd add it as the first commit of https://github.com/open-power-host-os/versions/pull/171, as its necessary to pass that build.

File: lib/repository.py [hunk 3]
```diff
@@ -63,7 +64,14 @@ def get_git_repository(remote_repo_url, parent_dir_path):
                       .format(name=name, url=remote_repo_url))
             repo.create_remote(MAIN_REMOTE_NAME, remote_repo_url)
         return repo
-    else:
+    except git.exc.InvalidGitRepositoryError:
+        # This exception will be thrown if the git repository has an
+        # invalid format, which can be due a previous svn repository
+        # at the repo_path directory.  In this case, we remove the old
+        # gcc repository and clone from scratch.
```
Comment (ghost) [inline lib/repository.py]: This could have other causes, I'm not sure mentioning gcc is necessary here. It could be mentioned in the commit message though. I suggest:
```This exception will be thrown if the Git repository has an invalid format, which could be due to the repository changing from SVN to Git, for example. The safer way to proceed is to remove the directory and re-clone from scratch.```

File: lib/repository.py [hunk 4]
```diff
@@ -63,7 +64,14 @@ def get_git_repository(remote_repo_url, parent_dir_path):
...
+        LOG.info("Repository path '%{repo_path}' has invalid format, removing "
+                 "and cloning from scratch".format(repo_path=repo_path))
+        shutil.rmtree(repo_path)
```
Comment (ghost) [inline lib/repository.py]: A `git.exc.NoSuchPathError` could also be raised if the path does not exist. We would have to execute the same clone action in this case.

To avoid code duplication, I suggest the following structure:
```python
repo = None
try:
  repo = GitRepository(repo_path)
except git.exc.InvalidGitRepositoryError:
  remove path
except git.exc.NoSuchPathError:
  pass

if repo is None:
  return GitRepository.clone_from
else:
  create main remote
  return repo
```

File: lib/iso_builder.py [hunk 5]
```diff
@@ -42,6 +42,8 @@ def __init__(self, config):
         (_, _, self.arch) = distro_utils.detect_distribution()
         self.mock_binary = self.common_config.get('mock_binary')
         self.mock_args = self.config.get('mock_args') or ""
+        self.pungi_binary = self.common_config.get('pungi_binary') or "pungi"
```
Comment (ghost) [inline lib/iso_builder.py:45]: I don't think this "default" should be here. I think we should fail if the value is not present in the config file or command line. You're just mimicking the behavior above though, so let's leave this discussion for another time. Alan's PR might fix this.

File: lib/config.py [hunk 6]
```diff
@@ -72,6 +72,12 @@
     ('--mock-args',):
         dict(help='Arguments passed to mock command', default=''),
 }
+PUNGI_ARGS = {
+    ('--pungi-binary',):
+        dict(help='Pungi binary path', default='/usr/bin/pungi'),
+    ('--pungi-args',):
+        dict(help='Arguments passed to pungi command', default=''),
```
Comment (ghost) [inline lib/config.py]: Those defaults will never be used, since the `config.yaml` file takes precedence. You may remove them. This is done for other options in Alan's PR.

**Review by ghost — CHANGES_REQUESTED** [review]
I'd squash all three commits pertaining to pungi args into one, but since they do not break anything if applied alone, you may do as you wish.

File: lib/repository.py [hunk 7]
```diff
@@ -48,9 +48,27 @@ def get_git_repository(remote_repo_url, parent_dir_path):
...
         repo = GitRepository(repo_path)
+    except git.exc.InvalidGitRepositoryError:
+        # This exception will be thrown if the Git repository has an
+        # invalid format, which could be due to the repository
+        # changing from SVN to Git, for example.  A way to proceed is
+        # removing the directory and cloning from scratch.
+        message = ("Repository path '{repo_path}' has invalid format. "
+                   "If you are unsure, remove it and try again."
+                   .format(repo_path=repo_path))
+        LOG.exception(message)
```
Comment (ghost) [inline lib/repository.py]: `LOG.exception` will print the stack trace. Since you're re-raising the exception, it should be printed anyway outside of here (possibly with the program abortion), so it's better to use `LOG.error` to avoid duplicating it.

File: lib/repository.py [hunk 8]
```diff
@@ -48,9 +48,27 @@ def get_git_repository(remote_repo_url, parent_dir_path):
...
+        raise
+    except git.exc.NoSuchPathError:
+        message = ("Repository path '{repo_path}' does not exist."
+                   .format(repo_path=repo_path))
+        LOG.exception(message)
```
Comment (ghost) [inline lib/repository.py]: I don't think it's a good idea to log with `ERROR` level within an expected code flow. I think `LOG.debug` would be enough, since the `clone_from` method already logs something to the user.

File: lib/repository.py [hunk 9]
```diff
@@ -48,9 +48,27 @@ def get_git_repository(remote_repo_url, parent_dir_path):
...
-        MAIN_REMOTE_NAME = "origin"
+    try:
         repo = GitRepository(repo_path)
+    except git.exc.InvalidGitRepositoryError:
+        # This exception will be thrown if the Git repository has an
+        # invalid format, which could be due to the repository
+        # changing from SVN to Git, for example.  A way to proceed is
+        # removing the directory and cloning from scratch.
+        message = ("Repository path '{repo_path}' has invalid format. "
+                   "If you are unsure, remove it and try again."
```
Comment (ghost) [inline lib/repository.py]: I'd change this slightly to clarify that the repository has an invalid format, not the path:
```
message = ("The Git repository in path '{repo_path}' has an invalid format. "
           "If you are unsure of what to do, remove this path and try again."
```

File: lib/repository.py (hunk 2 above)
Comment (mopsfelder) [reply lib/repository.py]: I've listed this pull as a dependency of https://github.com/open-power-host-os/versions/pull/171.  Good point!

File: lib/repository.py (hunk 1 above)
Comment (mopsfelder) [reply lib/repository.py]: Done.

File: lib/repository.py (hunk 3 above)
Comment (mopsfelder) [reply lib/repository.py]: Done.

File: lib/repository.py (hunk 4 above)
Comment (mopsfelder) [reply lib/repository.py]: Done.

File: lib/iso_builder.py (hunk 5 above)
Comment (mopsfelder) [inline lib/iso_builder.py]: I removed the `default=` from options.

File: lib/config.py (hunk 6 above)
Comment (mopsfelder) [reply lib/config.py]: I removed the `default=` from options.

File: lib/repository.py (hunk 9 above)
Comment (mopsfelder) [reply lib/repository.py]: Done.

File: lib/repository.py (hunk 7 above)
Comment (mopsfelder) [reply lib/repository.py]: Done.

File: lib/repository.py (hunk 8 above)
Comment (mopsfelder) [reply lib/repository.py]: Done.

**Review by farosas — APPROVED** [review]

**Review by ghost — COMMENTED** [review]
Just rebase on top of master for the test to pass.

**Review by ghost — APPROVED** [review]
